app/src/main/java/org/thoughtcrime/securesms/groups/GroupsV1MigrationUtil.java

```diff
@@ -0,0 +1,244 @@
+    }
+
+    if (!groupRecipient.isPushV1Group()) {
+      Log.w(TAG, "Not a V1 group!");
+      throw new InvalidMigrationStateException();
+    }
+
+    if (groupRecipient.getParticipants().size() > FeatureFlags.groupLimits().getHardLimit()) {
+      Log.w(TAG, "Too many members! Size: " + groupRecipient.getParticipants().size());
+      throw new InvalidMigrationStateException();
+    }
+
+    GroupId.V1     gv1Id        = groupRecipient.requireGroupId().requireV1();
+    GroupId.V2     gv2Id        = gv1Id.deriveV2MigrationGroupId();
+    GroupMasterKey gv2MasterKey = gv1Id.deriveV2MigrationMasterKey();
+    boolean        newlyCreated = false;
+
+    if (groupDatabase.groupExists(gv2Id)) {
+      Log.w(TAG, "We already have a V2 group for this V1 group! Must have been added before we were migration-capable.");
+      throw new InvalidMigrationStateException();
+    }
+
+    if (!groupRecipient.isActiveGroup()) {
+      Log.w(TAG, "Group is inactive! Can't migrate.");
+      throw new InvalidMigrationStateException();
+    }
+
+    switch (GroupManager.v2GroupStatus(context, gv2MasterKey)) {
+      case DOES_NOT_EXIST:
+        Log.i(TAG, "Group does not exist on the service.");
+
+        if (!groupRecipient.isProfileSharing()) {
+          Log.w(TAG, "Profile sharing is disabled! Can't migrate.");
+          throw new InvalidMigrationStateException();
+        }
+
+        if (!forced && SignalStore.internalValues().disableGv1AutoMigrateInitiation()) {
+          Log.w(TAG, "Auto migration initiation has been disabled! Skipping.");
+          throw new InvalidMigrationStateException();
+        }
+
+        List<Recipient> registeredMembers = RecipientUtil.getEligibleForSending(groupRecipient.getParticipants());
+
+        if (RecipientUtil.ensureUuidsAreAvailable(context, registeredMembers)) {
+          Log.i(TAG, "Newly-discovered UUIDs. Getting fresh recipients.");
+          registeredMembers = Stream.of(registeredMembers).map(Recipient::fresh).toList();
+        }
+
+        List<Recipient> possibleMembers = forced ? getMigratableManualMigrationMembers(registeredMembers)
+                                                 : getMigratableAutoMigrationMembers(registeredMembers);
+
+        if (!forced && !groupRecipient.hasName()) {
+          Log.w(TAG, "Group has no name. Skipping auto-migration.");
+          throw new InvalidMigrationStateException();
+        }
+
+        if (!forced && possibleMembers.size() != registeredMembers.size()) {
+          Log.w(TAG, "Not allowed to invite or leave registered users behind in an auto-migration! Skipping.");
+          throw new InvalidMigrationStateException();
+        }
+
+        Log.i(TAG, "Attempting to create group.");
+
+        try {
+          GroupManager.migrateGroupToServer(context, gv1Id, possibleMembers);
+          newlyCreated = true;
+          Log.i(TAG, "Successfully created!");
+        } catch (GroupChangeFailedException e) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: While wrapping the caught exception into a custom one, information about the caught exception is being lost, including information about the stack trace of the exception.

**Fix**: If the caught exception object does not contain sensitive information, consider passing it as the &#34;rootCause&#34; or inner exception parameter to the constructor of the new exception before throwing the new exception. (Note that not all exception constructors support inner exceptions. Use a wrapper exception that supports inner exceptions.) 
[Learn more](https://www.ibm.com/support/pages/best-practice-catching-and-re-throwing-java-exceptions)